webui/webui.go

```diff
@@ -0,0 +1,79 @@
+package webui
+
+import (
+	"bytes"
+	"context"
+	"embed"
+	"html/template"
+	"io/fs"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// This file exists to expose the Web UI components to the Bacalhau app.
+
+//go:embed build/**
+var buildFiles embed.FS
+
+// Return the index page for the web app with the passed API URL used. The app
+// running on each client will attempt to connect to the API URL to retrieve job
+// and node information.
+func IndexPage(apiURL string) ([]byte, error) {
+	rawIndex, err := buildFiles.ReadFile("build/index.html")
+	if err != nil {
+		return nil, err
+	}
+
+	indexTemplate, err := template.New("script").Parse(string(rawIndex))
+	if err != nil {
+		return nil, err
+	}
+
+	var output bytes.Buffer
+	err = indexTemplate.Execute(&output, struct{ BaseURL string }{apiURL})
+	return output.Bytes(), err
+}
+
+func ListenAndServe(ctx context.Context, baseURL string) error {
+	indexPage, err := IndexPage(baseURL)
+	if err != nil {
+		return err
+	}
+
+	files, err := fs.Sub(buildFiles, "build")
+	if err != nil {
+		return err
+	}
+
+	fileHandler := http.FileServer(http.FS(files))
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		match, err := fs.Glob(files, strings.TrimLeft(r.URL.Path, "/"))
+		if r.URL.Path == "/" || r.URL.Path == "/index.html" {
+			w.Write(indexPage)
+		} else if len(match) > 0 && err == nil {
+			// Serve a static file that matches the request path, if it exists
+			// We need to check this manually to enable the below behavior
+			fileHandler.ServeHTTP(w, r)
+		} else {
+			// If nothing matches, just serve the index page. This allows the
+			// app to modify the browser URL to refer to client side "pages"
+			// (e.g. `/jobs`) which the user can then successfully refresh,
+			// share or bookmark. If this wasn't here, the user would have to
+			// enter the app manually at `/` every time.
+			w.Write(indexPage)
+		}
+	})
+
+	server := &http.Server{
```

Comment:
With no setting for `Addr` here it will always be on port 80.  This is fine for now, but I suspect we'll need a little more flexibility later on (for instance when we support TLS for this) and so it's worth adding a ticket for a future change to add this (via the --web-ui parameter perhaps?).

Reply:
Will add this to configuration ticket (same as https://github.com/bacalhau-project/bacalhau/pull/3044#discussion_r1404243551)

Reply:
https://github.com/bacalhau-project/expanso-planning/issues/133